user/syscalls.S

```diff
@@ -0,0 +1,67 @@
+#include <mips/asm.h>
+#include <errno.h>
+
+    .set noreorder
+    .set noat
+
+/* This implementation uses newlib-specific errno processing. */
+#define DO_SYSCALL(num)  \
```

Comment:
What about that:

```asm
#define DO_SYSCALL(num)       \
    li $v0, num;              \
    syscall;                  \
    bltz $v0, syscall_failed; \
    nop;                      \
    jr $ra;                   \
    nop

SNESTED(syscall_failed, 24, $ra)
    addiu $sp, $sp, -24
    sw $ra, 20($sp)
    sw $v0, 16($sp)
    jal __errno
    nop
    lw $v1, 16($sp)
    negu $v1, $v1
    sw $v1, 0($v0)
    li $v0, -1
    lw $ra, 20($sp)
    jr $ra
    addiu $sp, $sp, 24
SEND(syscall_failed)
```

Please note that minimum stack frame size for non-leaf function is 24 according to [MIPS o32 Calling Convention Examples](http://cs.unm.edu/~jeffk/cs341f09/_media/o32callingconvention.pdf)